cadc-gms/src/main/java/org/opencadc/auth/StandardIdentityManager.java

```diff
@@ -0,0 +1,272 @@
+    }
+
+    @Override
+    public String toDisplayString(Subject subject) {
+        // prefer HttpPrincipal aka OIDC preferred_username for string output, eg logging
+        Set<HttpPrincipal> ps = subject.getPrincipals(HttpPrincipal.class);
+        if (!ps.isEmpty()) {
+            return ps.iterator().next().getName(); // kind of ugh
+        }
+        
+        // try X509
+        Set<X500Principal> px = subject.getPrincipals(X500Principal.class);
+        if (!px.isEmpty()) {
+            return px.iterator().next().getName(); // kind of ugh
+        }
+        
+        return null;
+    }
+    
+    private void validateOidcAccessToken(Subject s) {
+        log.debug("validateOidcAccessToken - START");
+        Set<AuthorizationTokenPrincipal> rawTokens = s.getPrincipals(AuthorizationTokenPrincipal.class);
+        
+        log.debug("token issuer: "  + oidcIssuer + " rawTokens: " + rawTokens.size());
+        if (oidcIssuer != null && !rawTokens.isEmpty()) {
+            URL u = getUserEndpoint();
+            for (AuthorizationTokenPrincipal raw : rawTokens) {
+                String credentials = null;
+                String challengeType = null;
+
+                // parse header
+                log.debug("header key: " + raw.getHeaderKey());
+                log.debug("header val: " + raw.getHeaderValue());
+                if (AuthenticationUtil.AUTHORIZATION_HEADER.equalsIgnoreCase(raw.getHeaderKey())) {
+                    String[] tval = raw.getHeaderValue().split(" ");
+                    if (tval.length == 2) {
+                        challengeType = tval[0];
+                        credentials = tval[1];
+                    } else {
+                        throw new NotAuthenticatedException(challengeType, NotAuthenticatedException.AuthError.INVALID_REQUEST,
+                            "invalid authorization");
+                    }
+                } // else: some other challenge 
+                log.debug("challenge type: " + challengeType);
+                log.debug("credentials: " + credentials);
+                    
+                // validate
+                if (challengeType != null && credentials != null) {
+                    try {
+                        HttpGet get = new HttpGet(u, true);
+                        get.setRequestProperty("authorization", raw.getHeaderValue());
+                        get.prepare();
+
+                        InputStream istream = get.getInputStream();
+                        String str = StringUtil.readFromInputStream(istream, "UTF-8");
+                        JSONObject json = new JSONObject(str);
+                        String sub = json.getString("sub");
+                        String username = json.getString("preferred_username");
+                        // TODO: register an X509 DN with IAM and see if I can get it back here
+
+                        NumericPrincipal np = new NumericPrincipal(UUID.fromString(sub));
```

Comment:
I wonder if it's okay to presume a UUID as the subject

Reply:
yes, `sub` is structurally a UUID, at least with IAM, although OIDC includes language in the spec that only issuer + sub is sure to be unique so it might not be a real (global) UUID. It's necessary to use NumericPrincipal. I think in future we could support int/long uids (like we do in AC) with a little more code, if necessary (eg if there is an OID provider impl that doesn't use UUID in sub).